### PR TITLE
TypeScript support

### PR DESCRIPTION
**Pull request by xg-wang**
~Document TypeScript usage as suggested by @DanielRosenwasser at https://github.com/ember-cli/ember-fetch/issues/72#issuecomment-365407346.~

~Provide a build time detection for `tsconfig.json` as a hint user needs the type support.~

This PR installs `ember-cli-typescript` and add typescript support doc to README.

Thanks @mike-north for the discussing the type resolution and detection strategy.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -1,9 +1,27 @@
-// Type definitions for Ember Fetch
+// Type definitions for ember-fetch
 // Project: https://github.com/ember-cli/ember-fetch
-// Definitions by: Toran Billups <https://github.com/toranb>
-// TypeScript Version: 2.3
+// Definitions by: Toran Billups <https://github.com/toranb>, Thomas Wang<https://github.com/xg-wang>
+/// <reference lib="dom"/>
 
 declare module 'fetch' {
```
Comment (mike-north) [inline index.d.ts]:
```suggestion
```

File: index.d.ts [hunk 2]
```diff
@@ -1,9 +1,27 @@
...
+    prototype: Response;
+    new(body?: BodyInit | null, init?: ResponseInit): Response;
+    error(): Response;
+    redirect(url: string, status?: number): Response;
+  };
+  export const AbortController: {
+    prototype: AbortController;
+    new(): AbortController;
+  };
+}
```
Comment (mike-north) [inline index.d.ts]:
```suggestion

```

File: index.d.ts [hunk 3]
```diff
@@ -1,9 +1,27 @@
-// Type definitions for Ember Fetch
+// Type definitions for ember-fetch
 // Project: https://github.com/ember-cli/ember-fetch
-// Definitions by: Toran Billups <https://github.com/toranb>
-// TypeScript Version: 2.3
+// Definitions by: Toran Billups <https://github.com/toranb>, Thomas Wang<https://github.com/xg-wang>
+/// <reference lib="dom"/>
```
Comment (mike-north) [inline index.d.ts]:
```suggestion
/// <reference lib="es2015" />
/// <reference lib="dom" />
```

**Comment by mike-north** [comment]
@xg-wang I filed a bug for the NPM issue we ran into https://npm.community/t/npm-install-for-package-with-local-dependency-fails/754/14

**Comment by xg-wang** [comment]
Great to know there's a pending PR in npm to fix the issue (which I'll document more about when this PR proceeds).
https://github.com/npm/cli/pull/3

File: addon/mixins/adapter-fetch.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  let headersObject: PlainObject = {};
+
+  if (headers) {
+    headers.forEach((value, key) => (headersObject[key] = value));
+  }
+
+  return headersObject;
+}
+
+export default Mixin.create({
```
Comment (xg-wang) [inline addon/mixins/adapter-fetch.ts]: I wanted to do `Mixin.create<TheMixinTypeLiteral, DS.RestAdapter>` here. But it seems it's not correct. For example `this.ajaxOptions` might be wrapped by base adapter's `this.ajaxOptions`'s `this._super(...arguments)` and return not necessarily `FetchOptions`.

**Comment by xg-wang** [comment]
@mike-north can you review again?

File: README.md [hunk 5]
```diff
@@ -35,6 +35,12 @@ Available imports:
 import fetch, { Headers, Request, Response, AbortController } from 'fetch';
 ```
 
+### Use with TypeScript
+To use `ember-fetch` with TypeScript or enable editor's type support, add `"fetch": "ember-cli/ember-fetch"` to your app's `devDependencies`.
+This will get the current state of `ember-fetch` from this GitHub repo as a dependency.
```
Comment (mike-north) [inline README.md]: I'm not sure we want to advise that people do this, since they'll end up tracking the master branch rather than SemVer versioned releases.

File: README.md [hunk 6]
```diff
@@ -35,6 +35,12 @@ Available imports:
...
+
+You can also add `"fetch": ["node_modules/ember-fetch"]` to your `tsconfig.json`.
```
Comment (mike-north) [inline README.md]: Please specify _where_ this should be added in `tsconfig.json`

File: addon/ajax.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+
+import fetch from 'fetch';
+
+export default async function ajax(
+  input: RequestInfo,
+  init?: RequestInit
+): Promise<Response> {
+  return fetch(input, init).then(response => {
```
Comment (mike-north) [inline addon/ajax.ts]: I would either

A) leave `ajax` as an `async` function, and use `await` to wait for promises instead of `.then`
B) treat `ajax` as a regular (non-`async`) function and use `.then` to wait for promises instead of `await`

Currently you're mixing (A) and (B) by making the function `async` but never `await`ing

File: addon/mixins/adapter-fetch.ts (hunk 4 above)
Comment (mike-north) [reply addon/mixins/adapter-fetch.ts]: from a typescript standpoint, there's no "clean" way to declare assumptions about the type of object the mixin will be applied to (this is one of the general pitfalls of using mixins in general).

The best you could do is explicitly specify `this` in your method signatures. For example

```ts
ajaxOptions(this: DataAdapter, url: string, type: Method, options: object): FetchOptions {
  // in here `this` will be a DataAdapter
  ...
} 

```

File: addon/mixins/adapter-fetch.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+   */
+  _fetchRequest(url: string, options: RequestInit): RSVP.Promise<Response> {
+    return fetch(url, options);
+  },
+
+  /**
+   * @override
+   */
+  ajaxSuccess(
+    adapter: any,
```
Comment (mike-north) [inline addon/mixins/adapter-fetch.ts]: Can we do better than `any` here? maybe a `DS.Adapter`?

File: addon/mixins/adapter-fetch.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    response: Response,
+    payload: Nullable<string | object>,
+    requestData: { url: string; method: string }
+  ): object | DS.AdapterError | RSVP.Promise<never> {
```
Comment (mike-north) [inline addon/mixins/adapter-fetch.ts]: Can we see of `RSVP.Promise<void>` works here?

File: addon/types.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import Mix from '@ember/polyfills/types';
+
+export type Nullable<T> = T | null | undefined;
+
+export interface PlainObject {
+  [key: string]: string | PlainObject | PlainObject[];
+}
+
+export interface PlainHeaders {
+  [key: string]: string | undefined | null;
```
Comment (mike-north) [inline addon/types.ts]:
```suggestion
  [key: string]?: string | null;
```

File: addon/types.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  headers?: PlainHeaders;
+};
+
+export type FetchOptions = Mix<
+  AjaxOptions,
+  { body?: BodyInit | null; method?: Method }
+>;
+
+export function isPlainObject(obj: any): obj is PlainObject {
+  return Object.prototype.toString.call(obj) === '[object Object]';
```
Comment (mike-north) [inline addon/types.ts]: this isn't really the right implementation for this guard, since it will pass for lots of objects that aren't `PlainObject`. For example `{ foo() { return 'bar'; } }` will still coerce to string as `[object Object]`

File: addon/utils/mung-options-for-fetch.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ */
+export default function mungOptionsForFetch(
+  options: AjaxOptions
+): FetchOptions {
+  const hash = assign(
+    {
+      credentials: 'same-origin'
+    },
+    options
+  ) as FetchOptions;
```
Comment (mike-north) [inline addon/utils/mung-options-for-fetch.ts]: Do you remember why we end up doing this, instead of adding `credentials` to the `FetchOptions` type?

File: addon/utils/serialize-query-params.ts [hunk 13]
```diff
@@ -1,15 +1,17 @@
...
  * This is directly copied from jQuery.param.
- * @param {Object} queryParamsObject
- * @returns {String}
  */
-export function serializeQueryParams(queryParamsObject) {
-  var s = [];
+export function serializeQueryParams(
+  queryParamsObject: object | string
+): string {
+  var s: any[] = [];
```
Comment (mike-north) [inline addon/utils/serialize-query-params.ts]:
```suggestion
  const s: any[] = [];
```

File: types/dummy/index.d.ts [hunk 14]
```diff
@@ -0,0 +1 @@
+
```
Comment (mike-north) [inline types/dummy/index.d.ts]: revert unnecessary whitespace

File: README.md (hunk 5 above)
Comment (xg-wang) [reply README.md]: I tend to document `tsconfig.json` only.

File: addon/ajax.ts (hunk 7 above)
Comment (xg-wang) [reply addon/ajax.ts]: oooh I was in the middle of transforming it to async but decide to leave it as is, because we are going to have a new [service](https://github.com/emberjs/ember.js/issues/17476) and probably deprecate this.

File: addon/mixins/adapter-fetch.ts (hunk 8 above)
Comment (xg-wang) [reply addon/mixins/adapter-fetch.ts]: > The best you could do is explicitly specify `this` in your method signatures.

We don't need this, `this` has a type here already
```js
static create<T, Base = EmberObject>(
    args?: MixinOrLiteral<T, Base> & ThisType<Fix<T & Base>>
): Mixin<T, Base>;
```

This is `Mixin.create` signature, so technically I can write `Mixin.create(FetchAdapter, DS.RESTAdapter>`

The issue is that

1. I have to manually define the object literal type
2. There might be **other** mixins, so type of `this` will still be inaccurate at runtime.

But I think it's still better than doing `adapter: any` here. Updated it to have an interface.

File: addon/mixins/adapter-fetch.ts (hunk 9 above)
Comment (xg-wang) [reply addon/mixins/adapter-fetch.ts]: `reject(reason?: any, label?: string | undefined): RSVP.Promise<never>`
It's typed for `RSVP.rejct`

File: addon/types.ts (hunk 10 above)
Comment (xg-wang) [reply addon/types.ts]: I think typescript doesn't let me do this.

File: addon/utils/mung-options-for-fetch.ts (hunk 12 above)
Comment (xg-wang) [reply addon/utils/mung-options-for-fetch.ts]: So I need to return a `FetchOptions` type required by ts, `assign` doesn't automatically gives me this kind of type. Explicit declaration is needed. `credentials` is an optional field in `fetch` API, but we 100% should add this to `FetchOptions` type. Updated.

File: addon/utils/serialize-query-params.ts (hunk 13 above)
Comment (xg-wang) [reply addon/utils/serialize-query-params.ts]: There're lots of `var` here because it was copy-pasted from jquery code. We can do a refactor in the future and leave this PR a type only one.

File: addon/types.ts (hunk 11 above)
Comment (xg-wang) [reply addon/types.ts]: I want to preserve all current behaviors so it be a type only PR. (I assume `Object.prototype.toString.call(obj)` equals `String(obj)`. There exists ways to break this but we can leave it in another fix PR.

File: addon/types.ts (hunk 10 above)
Comment (mike-north) [reply addon/types.ts]: you'll have to change it to a `type` instead of `interface`

```ts
interface Foo{
    bar: string;
    baz: number;
}

type A<T> = { [K in keyof T]?: T[K] | null }

type X = A<Foo>;/*
{
    bar?: string | null | undefined;
    baz?: number | null | undefined;
}
*/

```
https://typescript-play.js.org/#code/JYOwLgpgTgZghgYwgAgGIHt0G8BQz-IBGcUAXMgM5hSgDmA3HgcQF7kgCuAtodIwL44cYAJ4AHFAEEAPABUAfMgC8yLMgDaAaWShkAawgj0MZLIC6AfnKytZ5AB9knADbPkg4eJQANZchkY6PL0APQAVDi4BEQkVpTUdA5OHK5JHCAAJhAwoBAZjNGscZw80Ekubo7pWTkgeQI4YSE4QA

File: README.md (hunk 5 above)
Comment (mike-north) [reply README.md]: please clarify. This pertains to `package.json`, not `tsconfig.json`

> add `"fetch": "ember-cli/ember-fetch"` to your app's `devDependencies`.
This will get the current state of `ember-fetch` from this GitHub repo as a dependency.

File: addon/mixins/adapter-fetch.ts (hunk 9 above)
Comment (mike-north) [reply addon/mixins/adapter-fetch.ts]: `Promise<never>` is assignable to `Promise<void>` so you should be ok to make this change

```ts
function handlePromise(p: Promise<void>): void {

}
handlePromise(new Promise<never>(null as never))
```
https://typescript-play.js.org/#code/GYVwdgxgLglg9mABACwIZgCYBsCmAFAJzgFsYBnHACgAcAuRQk8nAHgDc4YMA+ASno5dEAbwBQogL6i0mXI1IVKYHAHcGRBa2VscBbkpBYsiVGUTbdvXkA

File: addon/types.ts (hunk 10 above)
Comment (xg-wang) [reply addon/types.ts]: I changed `PlainHeaders` and `PlainObject` from `interface` to `type`, and made `PlainObject` more generic.

File: README.md (hunk 5 above)
Comment (xg-wang) [reply README.md]: I removed the `"fetch": "ember-cli/ember-fetch"` instruction and only left a `tsconfig` example

**Comment by mike-north** [comment]
Looks like you've responded to all my feedback 👍. This looks good to me!
